R/utilities-units.R

```diff
@@ -336,3 +336,200 @@ initializeDimensionAndUnitLists <- function() {
+#'   available dimensions and units, see `ospsuite::ospDimensions` and
+#'   `ospsuite::ospUnits`, respectively.
+#'
+#' @seealso toUnit
+#'
+#' @examples
+#'
+#' # small dataframe to illustrate the conversion
+#' (df <- dplyr::tibble(
+#'   dataType = c(rep("simulated", 3), rep("observed", 3)),
+#'   xValues = c(0, 14.482, 28.965, 0, 1, 2),
+#'   xUnit = "min", xDimension = "Time",
+#'   yValues = c(1, 1, 1, 1, 1, 1),
+#'   yUnit = c("mol", "mol", "mol", "g", "g", "g"),
+#'   yDimension = c("Amount", "Amount", "Amount", "Mass", "Mass", "Mass"),
+#'   yErrorValues = c(2.747, 2.918, 2.746, NA, NA, NA),
+#'   yErrorUnit = c("mol", "mol", "mol", "g", "g", "g"),
+#'   molWeight = c(10, 10, 20, 20, 20, 10)
+#' ))
+#'
+#' # default conversion
+#' ospsuite:::.unitConverter(df)
+#'
+#' # customizing conversion with specified unit(s)
+#' ospsuite:::.unitConverter(df, xUnit = ospUnits$Time$h)
+#' ospsuite:::.unitConverter(df, yUnit = ospUnits$Mass$kg)
+#' ospsuite:::.unitConverter(df, xUnit = ospUnits$Time$s, yUnit = ospUnits$Amount$mmol)
+#'
+#' @keywords internal
+.unitConverter <- function(data, xUnit = NULL, yUnit = NULL) {
+  # target units --------------------------
+
+  # No validation of inputs for this non-exported function.
+  # All validation will take place in the `DataCombined` class itself.
+
+  # The observed and simulated data should have the same units for
+  # visual/graphical comparison.
+  #
+  # Therefore, if target units are not specified by the user, we need to choose
+  # one ourselves. For no special reason, the first element from a vector of
+  # unique units will be selected: one for X-axis, and one for Y-axis, i.e.
+  xTargetUnit <- xUnit %||% unique(data$xUnit)[[1]]
+  yTargetUnit <- yUnit %||% unique(data$yUnit)[[1]]
+
+  # internal --------------------------
+
+  # The strategy is to split the data frame for each source unit and carry out
+  # conversion separately per data frame. This is the most performant options
+  # since there can only be as many expensive calls to `toUnit()`/`rClr` as
+  # there are source units.
+  #
+  # The problem occurs when source units are missing (`NA`). The `toUnit`
+  # function can handle them but not `split()`, which would drop the entire
+  # section of the data frame corresponding to `NA` and thus there will be loss
+  # of data when a data frame is split into a list of data frames.
+  #
+  # The trick is to create copies of source unit and molecular weight columns
+  # and fill in the missing values with something other than `NA`. Using these
+  # new columns with `split()` makes sure that parts of a data frame where
+  # source units are missing won't be dropped. Note that the original columns
+  # containing source units remain untouched.
+  #
+  # These newly created columns are removed before the converted data frame is
+  # returned to the user.
+  data <- data %>%
+    dplyr::mutate(
+      dplyr::across(
+        .cols = dplyr::matches("Unit$|Weight$"),
```

Comment:
Do you know for sure that those columns will be here? 
Can you add a comment explaining what columns you are targeting with this ?

Reply:
> Do you know for sure that those columns will be here?

Yes. `DataCombined$toDataFrame()` guarantees that. Also, the pattern matching makes sure that nothing will happen even if they were not to be found. But, in that case, I am not sure why would be using this internal function any way.

> Can you add a comment explaining what columns you are targeting with this ?

Done.